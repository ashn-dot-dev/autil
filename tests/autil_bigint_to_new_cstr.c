#define AUTIL_IMPLEMENTATION
#include "../autil.h"
#include "test.h"

static void
test_valid(char const* input, char const* fmt)
{
    assert(input != NULL);

    struct autil_bigint* const num = autil_bigint_new_cstr(input);
    ASSERT(num != NULL);

    char* const cstr = autil_bigint_to_new_cstr(num, fmt);
    ASSERT(cstr != NULL);

    printf("%-26s | %-8s | \"%s\"\n", input, fmt ? fmt : "NULL", cstr);

    autil_bigint_del(num);
    autil_xalloc(cstr, AUTIL_XALLOC_FREE);
}

int
main(void)
{
    test_valid("+123456789123456789", NULL);
    test_valid("-123456789123456789", NULL);

    fputc('\n', stdout);
    test_valid("+123456789123456789", "d");
    test_valid("-123456789123456789", "d");

    fputc('\n', stdout);
    // Without padding these should do nothing.
    test_valid("+123456789123456789", "#d");
    test_valid("-123456789123456789", "#d");
    test_valid("+123456789123456789", "0d");
    test_valid("-123456789123456789", "0d");
    test_valid("+123456789123456789", " d");
    test_valid("-123456789123456789", " d");

    fputc('\n', stdout);
    test_valid("+123456789123456789", "+d");
    test_valid("-123456789123456789", "+d");
    test_valid("+123456789123456789", " d");
    test_valid("-123456789123456789", " d");

    fputc('\n', stdout);
    test_valid("+123456789123456789", "24d");
    test_valid("-123456789123456789", "24d");
    test_valid("+123456789123456789", "-24d");
    test_valid("-123456789123456789", "-24d");
    test_valid("+123456789123456789", "024d");
    test_valid("-123456789123456789", "024d");
    test_valid("+123456789123456789", "0-24d");
    test_valid("-123456789123456789", "0-24d");
    test_valid("+123456789123456789", " 24d");
    test_valid("-123456789123456789", " 24d");
    test_valid("+123456789123456789", " -24d");
    test_valid("-123456789123456789", " -24d");

    fputc('\n', stdout);
    test_valid("0b111100001100110010101010", "b");
    test_valid("0b111100001100110010101010", "#b");

    fputc('\n', stdout);
    test_valid("0o765432107654321076543210", "o");
    test_valid("0o765432107654321076543210", "#o");

    fputc('\n', stdout);
    test_valid("0xbeefbeefbeefbeef", "x");
    test_valid("0xbeefbeefbeefbeef", "X");
    test_valid("0xbeefbeefbeefbeef", "#x");
    test_valid("0xbeefbeefbeefbeef", "#X");

    struct autil_bigint* const num = autil_bigint_new_cstr("0xDEADBEEF");
    EXPECT(NULL == autil_bigint_to_new_cstr(num, ""));
    EXPECT(NULL == autil_bigint_to_new_cstr(num, "#"));
    EXPECT(NULL == autil_bigint_to_new_cstr(num, "0"));
    EXPECT(NULL == autil_bigint_to_new_cstr(num, "+"));
    EXPECT(NULL == autil_bigint_to_new_cstr(num, "-"));
    EXPECT(NULL == autil_bigint_to_new_cstr(num, " "));
    EXPECT(NULL == autil_bigint_to_new_cstr(num, "1"));
    autil_bigint_del(num);

    return EXIT_SUCCESS;
}
