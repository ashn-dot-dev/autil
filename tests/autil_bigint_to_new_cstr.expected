+123456789123456789        | NULL     | "123456789123456789"
-123456789123456789        | NULL     | "-123456789123456789"

+123456789123456789        | d        | "123456789123456789"
-123456789123456789        | d        | "-123456789123456789"

+123456789123456789        | #d       | "123456789123456789"
-123456789123456789        | #d       | "-123456789123456789"
+123456789123456789        | 0d       | "123456789123456789"
-123456789123456789        | 0d       | "-123456789123456789"
+123456789123456789        |  d       | " 123456789123456789"
-123456789123456789        |  d       | "-123456789123456789"

+123456789123456789        | +d       | "+123456789123456789"
-123456789123456789        | +d       | "-123456789123456789"
+123456789123456789        |  d       | " 123456789123456789"
-123456789123456789        |  d       | "-123456789123456789"

+123456789123456789        | 24d      | "      123456789123456789"
-123456789123456789        | 24d      | "     -123456789123456789"
+123456789123456789        | -24d     | "123456789123456789      "
-123456789123456789        | -24d     | "-123456789123456789     "
+123456789123456789        | 024d     | "000000123456789123456789"
-123456789123456789        | 024d     | "-00000123456789123456789"
+123456789123456789        | 0-24d    | "123456789123456789      "
-123456789123456789        | 0-24d    | "-123456789123456789     "
+123456789123456789        |  24d     | "      123456789123456789"
-123456789123456789        |  24d     | "     -123456789123456789"
+123456789123456789        |  -24d    | " 123456789123456789     "
-123456789123456789        |  -24d    | "-123456789123456789     "

0b111100001100110010101010 | b        | "111100001100110010101010"
0b111100001100110010101010 | #b       | "0b111100001100110010101010"

0o765432107654321076543210 | o        | "765432107654321076543210"
0o765432107654321076543210 | #o       | "0o765432107654321076543210"

0xbeefbeefbeefbeef         | x        | "beefbeefbeefbeef"
0xbeefbeefbeefbeef         | X        | "BEEFBEEFBEEFBEEF"
0xbeefbeefbeefbeef         | #x       | "0xbeefbeefbeefbeef"
0xbeefbeefbeefbeef         | #X       | "0xBEEFBEEFBEEFBEEF"
